### PR TITLE
add pytorch pkg to mariner

### PR DESCRIPTION
**Pull request by mandeepsplaha**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [X] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [X] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
This PR adds `pytorch` package to Mariner 2.0 to be used in building Pytorch Golden container.

###### Change Log  

Adding pytorch package to Mariner 2.0

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- ARM64 Buddy Build Id: 306737
- AMD64 Buddy Build Id: 306736

### CODE REVIEW

File: SPECS/LICENSES-AND-NOTICES/data/licenses.json
Comment (oliviacrain) [inline SPECS/LICENSES-AND-NOTICES/data/licenses.json:2323]: I'm not too familiar with openEuler- I managed to find [their spec](https://gitee.com/src-openeuler/pytorch/blob/master/pytorch.spec), but I couldn't find any information on how they license the spec files they create. Mind pointing me to some documentation on the subject?

File: SPECS/python-pytorch/python-pytorch.spec
Comment (oliviacrain) [inline SPECS/python-pytorch/python-pytorch.spec:23]: We don't have automatic generation of runtime requirements for python specs like a lot of other distros do. There seems to be some runtime requirements missing, judging by [upstream's `setup.py`](https://github.com/pytorch/pytorch/blob/master/setup.py#L1015-L1020) We should add those in manually.

**Review by oliviacrain — CHANGES_REQUESTED** [review]

**Comment by anphel31** [comment]
Can you update the PR title to remove your username?

File: SPECS/python-pytorch/python-pytorch.spec
Comment (mandeepsplaha) [reply SPECS/python-pytorch/python-pytorch.spec]: We do have `filelock` and `typing_extensions` packages so I added those two as **Requires**, but we don't have `networkx` and `sympy` in Mariner. Interestingly, the cpu-only version of the [upstream pytorch image](https://hub.docker.com/r/bitnami/pytorch) that we are building our pytorch cpu-only image after doesn't seem to have `networkx` or `sympy` included. The image with `cuda` however, has `networkx` so we can definitely include it in next iteration when we build pytorch with cuda.

File: SPECS/LICENSES-AND-NOTICES/data/licenses.json
Comment (mandeepsplaha) [reply SPECS/LICENSES-AND-NOTICES/data/licenses.json]: After talking to CELA, we have decided to take the OpenEuler's PyTorch spec file under the PyTorch license (i.e., BSD-3), since there's no specific license found in OpenEuler's packages' repo.

File: SPECS/python-pytorch/python-pytorch.spec
Comment (mandeepsplaha) [reply SPECS/python-pytorch/python-pytorch.spec]: I tried building pytorch with `pyproject-rpm-macros` but I failed to get it past the `%pyproject_save_files` step. Since keeping this implementation doesn't break functionality, I will defer using `pyproject-rpm-macros` to next time when we come back to this spec. I have created a bug in ADO (ID: 43259856) to track this change.

File: cgmanifest.json
Comment (mandeepsplaha) [reply cgmanifest.json:23806]: Renamed the spec file to pytorch so this change is no longer necessary here.

File: SPECS/python-pytorch/python-pytorch.spec
Comment (mandeepsplaha) [reply SPECS/python-pytorch/python-pytorch.spec]: Renamed the spec file from `python-pytorch.spec` to `pytorch.spec`.

File: SPECS/python-pytorch/python-pytorch.spec
Comment (mandeepsplaha) [reply SPECS/python-pytorch/python-pytorch.spec]: Updated.

File: SPECS/python-pytorch/python-pytorch.spec
Comment (mandeepsplaha) [reply SPECS/python-pytorch/python-pytorch.spec]: Updated.

File: toolkit/scripts/spec_source_attributions.py
Comment (mandeepsplaha) [reply toolkit/scripts/spec_source_attributions.py]: Done.

File: SPECS/pytorch/pytorch.signatures.json
Comment (osamaesmailmsft) [inline SPECS/pytorch/pytorch.signatures.json:4]: Question: Why generate two separate tarballs for the submodules and pytorch itself? Is it for modularity, like not having to update the whole tarball when a submodule changes? The reason I ask is for future reference / consistency.

File: SPECS/pytorch/pytorch.spec
Comment (osamaesmailmsft) [inline SPECS/pytorch/pytorch.spec:59]: Do we not support CUDA? From my experience, CUDA is a massive speedup for larger training sets, so just asking because I'm not sure what our configuration and such is.

File: SPECS/pytorch/pytorch.signatures.json
Comment (mandeepsplaha) [reply SPECS/pytorch/pytorch.signatures.json:4]: No strong reason other than a couple of example packages with similar pattern: `ltp` and `influxdb`. I suppose we could combine the submodules with pytorch itself. I'd defer to the next time we touch this spec.

Comment (PawelWMS) [reply SPECS/pytorch/pytorch.signatures.json:4]: Our upgrade automation relies on having a clean tarball from upstream and the `generate_source_tarball.sh` script to download and prepare the submodules. So that's another reason we're keeping these separate.

Also, a clean upstream tarball makes it possible to verify we've got the right file by comparing our signature with the downloaded upstream file.

File: SPECS/LICENSES-AND-NOTICES/data/licenses.json
Comment (PawelWMS) [inline SPECS/LICENSES-AND-NOTICES/data/licenses.json:2319]: How do we know they release their specs under this license? I tried a brief search and could find [their repo for PyTorch](https://gitee.com/src-openeuler/pytorch), but no licensing info, so I got curious.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:11]: For the sake of following what we try to do in most specs:

```suggestion
Source0:        https://github.com/pytorch/pytorch/releases/download/v%{version}/%{name}-v%{version}.tar.gz#/%{name}-%{version}.tar.gz
```

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:36]: Let's move all the `BuildRequires` to right below the sources to follow our spec writing conventions.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:44]: I'd re-name that to `doc` - that's more in line, with the rest of our packages.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:46]:
```suggestion
Provides:       python3-pytorch-doc
Provides:       pytorch-doc
```

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:36]: Suggestion: to be consistent with the naming, I'd suggest sticking to the word `pytorch` instead of plain `torch`.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:46]: Let's be consistent and whatever package name and `Provides` we use for `python3-pytorch`, let's duplicate them here with the `-doc` suffix.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:56]: Wouldn't doing that in one line work as well:
```suggestion
%autosetup -a 1
```

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:69]: Suggestion: let's combine these lines to make it easier to see they do the same thing for different directories:
```suggestion
for directory doc docs example examples; do
  if [ -d $directory ]; then
    cp -arf $directory %{buildroot}/%{_pkgdocdir}
  fi
done
```

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:89]: Suggestions:
- Let's use loops to make it clear which sections do the same thing just for different directories.
- Let's avoid explicit use of `usr/` and rely on `%{_prefix}` instead. This will also allow us to drop `push` and the additional `mv` at the end.

```suggestion
for directory in doc docs example examples; do
  if [ -d $directory ]; then
    cp -arf $directory %{buildroot}/%{_pkgdocdir}
  fi
done

for directory in lib lib64 bin sbin; do
  if [ -d $directory ]; then
    find %{buildroot}%{_prefix}/$directory -type f -printf "/%h/%f\n" >> filelist.lst
  fi
done

if [ -d usr/share/man ]; then
    find %{buildroot}%{_prefix}/share/man -type f -printf "/%h/%f.gz\n" >> doclist.lst
fi
```

**Review by PawelWMS — CHANGES_REQUESTED** [review]
A few small clean-up requests and a question about the license.

File: SPECS/LICENSES-AND-NOTICES/data/licenses.json
Comment (mandeepsplaha) [reply SPECS/LICENSES-AND-NOTICES/data/licenses.json:2319]: We talked to CELA about not finding a license in openEuler's repo for this package and their advice is to take it under PyTorch license.

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:80]: My bad, sorry:

```suggestion
for directory in %{buildroot}%{_prefix}/{lib,lib64,bin,sbin}; do
  if [ -d $directory ]; then
    find $directory -type f -printf "/%h/%f\n" >> filelist.lst
  fi
done

man_dir="%{buildroot}%{_prefix}/share/man"
if [ -d $man_dir ]; then
    find $man_dir -type f -printf "/%h/%f.gz\n" >> doclist.lst
fi
```

File: SPECS/pytorch/pytorch.spec
Comment (rikenm1) [reply SPECS/pytorch/pytorch.spec]: @PawelWMS this package is called python3-torch in all other linux distros. 
Even fact even the python module that gets imported is [torch](https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html). 

Should we move away from the convention?

File: SPECS/pytorch/pytorch.spec
Comment (PawelWMS) [inline SPECS/pytorch/pytorch.spec:13]: I'd only move this back to the top of the spec before merging.

**Review by PawelWMS — APPROVED** [review]
One more linting request but looks good!

File: SPECS/pytorch/pytorch.spec
Comment (oliviacrain) [inline SPECS/pytorch/pytorch.spec:67]: License needs to be packaged here, and in the doc subpackage as well

**Review by oliviacrain — CHANGES_REQUESTED** [review]

File: SPECS/pytorch/pytorch.spec
Comment (mandeepsplaha) [reply SPECS/pytorch/pytorch.spec:67]: Done. Ran a local build after this change to confirm the package built successfully.

**Review by oliviacrain — APPROVED** [review]

**Pull request merged**
